controllers/osbuildenvconfig_controller.go

```diff
@@ -47,16 +132,417 @@ type OSBuildEnvConfigReconciler struct {
+			return r.Finalize(ctx, reqLogger, instance)
+		}
+		return ctrl.Result{}, nil
+	}
+	// resource is alive
+	return r.Update(ctx, reqLogger, instance)
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *OSBuildEnvConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&osbuilderprojectflottaiov1alpha1.OSBuildEnvConfig{}).
+		For(&osbuildv1alpha1.OSBuildEnvConfig{}).
 		Complete(r)
 }
+
+func (r *OSBuildEnvConfigReconciler) Update(ctx context.Context, reqLogger logr.Logger, instance *osbuildv1alpha1.OSBuildEnvConfig) (ctrl.Result, error) {
+	reqLogger.Info("Updating state for OSBuildEnvConfig",
+		"name", instance.Name,
+		"UID", instance.UID,
+	)
+
+	changed, err := r.addFinalizer(ctx, instance)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	} else if changed {
+		reqLogger.Info("Added finalizer")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	_, changed, err = r.getOrCreateCertificate(
+		ctx,
+		reqLogger,
+		instance,
+		composerCertificateName,
+		[]string{
+			composerComposerAPIServiceName,
+			composerWorkerAPIServiceName,
+		},
+	)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	} else if changed {
+		reqLogger.Info("Created composer certificate")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	composerDeploymentParams := composerDeploymentParameters{
+		ComposerDeploymentNamespace:      conf.GlobalConf.WorkingNamespace,
+		ComposerDeploymentName:           composerDeploymentName,
+		ComposerImageName:                "quay.io/app-sre/composer",
+		ComposerImageTag:                 "fc87b17",
+		ProxyImageName:                   "docker.io/envoyproxy/envoy",
+		ProxyImageTag:                    "v1.21-latest",
+		ComposerAPIInternalPort:          18080,
+		ComposerAPIExternalPort:          8080,
+		WorkerAPIInternalPort:            18700,
+		WorkerAPIExternalPort:            8700,
+		PgSSLMode:                        pgSSLModeDefault,
+		ProxyCertsDir:                    "/etc/certs",
+		PGSQLSecretName:                  "",
+		ComposerConfigMapName:            composerConfigMapName,
+		ComposerCertsSecretName:          composerCertificateName,
+		ComposerCertsSecretPublicCertKey: certificateSecretCrtKey,
+		ComposerCertsSecretPrivateKeyKey: certificateSecretKeyKey,
+		ComposerCertsSecretCACertKey:     certificateSecretCACrtKey,
+		ProxyConfigMapName:               composerProxyConfigMapName,
+	}
+
+	if instance.Spec.Composer != nil && instance.Spec.Composer.PSQL != nil {
+		composerDeploymentParams.PGSQLSecretName = instance.Spec.Composer.PSQL.ConnectionSecretReference.Name
+		if instance.Spec.Composer.PSQL.SSLMode != nil {
+			composerDeploymentParams.PgSSLMode = string(*instance.Spec.Composer.PSQL.SSLMode)
+		}
+	} else {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("creating a PSQL service is not yet implemented")
+	}
+
+	_, changed, err = r.getOrCreateComposerConfigMap(ctx, instance)
```

Comment:
Each step of the `Update` will be executed in separate reconciliation - did you consider creating as many resources as possible during one reconciliation? That would reduce time needed to create an environment.
The code now looks clean and easy to understand - would that be severely impacted by that kind of change?

Reply:
I don't have strong opinion here. I think this pattern would work good (better than all in one go) if accompanied by status conditions reflecting progress. Do you have plans for adding them?

Reply:
Yes, I am planning on adding progress events and update the conditions in the status